crates/red_knot_python_semantic/src/types.rs

```diff
@@ -145,8 +145,27 @@ impl<'db> Type<'db> {
         matches!(self, Type::Unbound)
     }
 
-    pub const fn is_unknown(&self) -> bool {
-        matches!(self, Type::Unknown)
+    pub fn may_be_unbound(&self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Unbound => true,
+            Type::Union(union) => union.contains(db, Type::Unbound),
+            _ => false,
+        }
+    }
+
+    #[must_use]
+    pub fn replace_unbound_with(&self, db: &'db dyn Db, replacement: Type<'db>) -> Type<'db> {
+        match self {
+            Type::Unbound => replacement,
+            Type::Union(union) => union
```

Comment:
Probably a stupid question but: What's the reason that we only do this for unions and not intersections?

Reply:
It's a great question, and it actually highlights a change I should just make in this PR.

Unbound should never appear in an intersection because if it did, the intersection should simplify to just Unbound. A symbol that must be both Unbound and some other type, is just unbound, the other type constraint is irrelevant.

But I should add a comment to that effect, and also our `IntersectionBuilder` doesn't actually do this yet, so I should add that in this PR as well.